packages/smart-router/evm/v3-router/functions/getPairCombinations.ts

```diff
@@ -1,12 +1,87 @@
 import { ChainId } from '@pancakeswap/chains'
 import { Currency, Token } from '@pancakeswap/sdk'
+import { getTokensByChain } from '@pancakeswap/tokens'
+import type { Address } from 'viem'
+import { GAUGES_CONFIG, GaugeConfig, GaugeType } from '@pancakeswap/gauges'
 import flatMap from 'lodash/flatMap.js'
 import memoize from 'lodash/memoize.js'
+import uniqBy from 'lodash/uniqBy.js'
 
 import { ADDITIONAL_BASES, BASES_TO_CHECK_TRADES_AGAINST, CUSTOM_BASES } from '../../constants'
 import { wrappedCurrency } from '../../utils/currency'
 import { isCurrenciesSameChain } from '../utils'
 
+const allGuages = GAUGES_CONFIG[ChainId.BSC]
+
+const getToken = memoize(
+  (chainId?: ChainId, address?: Address): Token | undefined => {
+    if (!chainId || !address) {
+      return undefined
+    }
+    const tokens = getTokensByChain(chainId)
+    for (const token of tokens) {
+      if (token.address.toLowerCase() === address.toLowerCase()) {
+        return token
+      }
+    }
+    return undefined
+  },
+  (chainId, address) => `${chainId}_${address}`,
+)
+
+// TODO: move to gauges
+const getGaugesByChain = memoize(
+  (chainId?: ChainId): GaugeConfig[] => allGuages.filter((gauge) => gauge.chainId === chainId),
+  (chainId) => chainId,
+)
+
+function isTokenInCommonBases(token?: Token) {
+  return Boolean(token && BASES_TO_CHECK_TRADES_AGAINST[token.chainId as ChainId].find((t) => t.equals(token)))
+}
+
+const getTokenBasesFromGauges = memoize(
+  (currency?: Currency): Token[] => {
+    const chainId: ChainId | undefined = currency?.chainId
+    const address = currency?.wrapped.address
+    const gauges = getGaugesByChain(currency?.chainId)
+    const bases = new Set<Token>()
+    const addTokenToBases = (token?: Token) => token && !isTokenInCommonBases(token) && bases.add(token)
+    const addTokensToBases = (tokens: Token[]) => tokens.forEach(addTokenToBases)
```

Comment:
Are you trying to use `Set` to dedupe `Token`? I guess it won't dedupe even it's same Token

Reply:
The `getToken` function is memoized so for the same token we're going to add the same reference. Should be good?